src/meta/src/rpc/service/cloud_service.rs

```diff
@@ -0,0 +1,184 @@
+        if let Some(connection_name) = source_cfg.get("connection.name") {
+            let connection = self
+                .catalog_manager
+                .get_connection_by_name(connection_name)
+                .await;
+            if let Err(e) = connection {
+                return Ok(new_rw_cloud_validate_source_err(
+                    ErrorType::PrivatelinkConnectionNotFound,
+                    e.to_string(),
+                ));
+            }
+            if let Some(info) = connection.unwrap().info {
+                match info {
+                    PrivateLinkService(service) => {
+                        let privatelink_status = cli
+                            .is_vpc_endpoint_ready(service.endpoint_id.as_str())
+                            .await;
+                        if let Err(e) = privatelink_status {
+                            return Ok(new_rw_cloud_validate_source_err(
+                                ErrorType::PrivatelinkUnavailable,
+                                e.to_string(),
+                            ));
+                        }
+                        if !privatelink_status.unwrap() {
+                            return Ok(new_rw_cloud_validate_source_err(
+                                ErrorType::PrivatelinkUnavailable,
+                                format!(
+                                    "Private link endpoint {} is not ready",
+                                    service.endpoint_id,
+                                ),
+                            ));
+                        }
+                        if let Err(e) =
+                            insert_privatelink_broker_rewrite_map(&service, &mut source_cfg)
+                        {
+                            return Ok(new_rw_cloud_validate_source_err(
+                                ErrorType::PrivatelinkResolveErr,
+                                e.to_string(),
+                            ));
+                        }
+                    }
+                }
+            } else {
+                return Ok(new_rw_cloud_validate_source_err(
+                    ErrorType::PrivatelinkResolveErr,
+                    format!("connection {} has no info available", connection_name),
+                ));
+            }
+        }
+        // try fetch kafka metadata, return error message on failure
+        let source_cfg: HashMap<String, String> =
+            source_cfg.into_iter().map(|(k, v)| (k, v)).collect();
+        let props = ConnectorProperties::extract(source_cfg);
+        if let Err(e) = props {
+            return Ok(new_rw_cloud_validate_source_err(
+                ErrorType::KafkaInvalidProperties,
+                e.to_string(),
+            ));
+        };
+        let enumerator = SplitEnumeratorImpl::create(props.unwrap()).await;
+        if let Err(e) = enumerator {
+            return Ok(Response::new(RwCloudValidateSourceResponse {
```

Comment:
Why don't use the function to construct response?

Reply:
😅